### DISCUSSION TITLE
Should I be concerned every single SchedulerJob fails?

### DISCUSSION THREAD

**Discussion by notatallshaw-gts** [apache/airflow#28418] [Q&A]
Looking at my Jobs table I noticed that SchedulerJob runs infrequently but fails every time.

I think the pattern is that it starts running, never ends, and then when the scheduler is restart the new scheduler marks the old one as failed.

Is this an expected pattern or should SchedulerJob complete?

**Comment by pingzh** [comment]
i don't think it should happen often. did you check the scheduler log? also what's your airflow version?

**Reply by notatallshaw-gts** [reply]
No error in scheduler logs, Airflow 2.3.4. My hypothesis is when we shut down the docker container it doesn't have chance to finish shutting down before it's killed, so I'm going to add a longer time on the docker stop command and see if that helps.

**Reply by potiuk** [reply]
You should likely to the same thing that other deployments do. Usually when you kill a running process you escalate signals that are sent to the process: TERM (wait graceful period) -> TERM (wait a bit - give it time to stop) ->  HUP (wait a bit again) -> KILL (as a last resort). This is what for example K8S does by default when stopping containers.

Also I think you should possibly look at signal propagation in your image. Not sure which docker images you use but there are different termination behaviours depending on entrypoing you use. You need to have `init` process started in your container in order to properly propagate signals. If you have "bash" script as entrypoint (which many people mistakenly do) the signals will not be propagated properly and depending how you run airflow, the TERM, HUP signals might be silently swallowed by the bash process running as init. You can either use `--init` flag to start  tiny when you start your container or embed `tini` (minimal init process that does what init process should do) or use another init process (we use `dumb-init` in Airlfow PROD image.

Also depending on the process you might want to propagate the signals to single airflow process or the whole airflow process group - airflow scheduler starts a number of forked processes in the same process group and propagating signals to all of them is the fastest and more certain way to kill all of them - this is what dumb-init does by default.

However if you run celery you want to only send signal to main process because celery handles signals on their own and if you want to do a graceful shutdown, you send TERM first, then it will put the celery worker in offline state, then you wait enough time for the running tasks to complete (you need to give it enough time - longer that longest running task) and then second TERM will stop all the running processes that celery spawns.  You can read about it here for Airlfow PROD image: https://airflow.apache.org/docs/docker-stack/entrypoint.html#signal-propagation
